config/internal/merge/merge_test.go

```diff
@@ -0,0 +1,516 @@
+			},
+		},
+		{
+			name: "overwrite(bool)",
+			dst: &Example{
+				BoolField: true,
+			},
+			src: &map[string]any{
+				"bool_field": false,
+			},
+			expected: &Example{
+				BoolField: false,
+			},
+		},
+		{
+			name: "dst(string)",
+			dst:  &Example{},
+			src: &map[string]any{
+				"string_field": "some string",
+			},
+			expected: &Example{
+				StringField: "some string",
+			},
+		},
+		{
+			name: "src(string)",
+			dst: &Example{
+				StringField: "some string",
+			},
+			src: &map[string]any{},
+			expected: &Example{
+				StringField: "some string",
+			},
+		},
+		{
+			name: "overwrite(string)",
+			dst: &Example{
+				StringField: "some string",
+			},
+			src: &map[string]any{
+				"string_field": "other string",
+			},
+			expected: &Example{
+				StringField: "other string",
+			},
+		},
+		{
+			name: "dst(slice)",
+			dst:  &Example{},
+			src: &map[string]any{
+				"slice": []any{"some", "strings"},
+			},
+			expected: &Example{
+				Slice: []string{"some", "strings"},
+			},
+		},
+		{
+			name: "src(slice)",
+			dst: &Example{
+				Slice: []string{"some", "strings"},
+			},
+			src: &map[string]any{},
+			expected: &Example{
+				Slice: []string{"some", "strings"},
+			},
+		},
+		{
+			name: "overwrite(slice)",
+			dst: &Example{
+				Slice: []string{"some", "strings"},
+			},
+			src: &map[string]any{
+				"slice": []any{"other", "strings"},
+			},
+			expected: &Example{
+				Slice: []string{"other", "strings"},
+			},
+		},
+		{
+			name: "src(map)",
+			dst:  &Example{},
+			src: &map[string]any{
+				"map": map[any]any{
+					"some": "map",
+				},
+			},
+			expected: &Example{
+				Map: map[any]any{
+					"some": "map",
+				},
+			},
+		},
+		{
+			name: "dst(map)",
+			dst: &Example{
+				Map: map[any]any{
+					"some": "map",
+				},
+			},
+			src: &map[string]any{},
+			expected: &Example{
+				Map: map[any]any{
+					"some": "map",
+				},
+			},
+		},
+		{
+			name: "overwrite(map)",
+			dst: &Example{
+				Map: map[any]any{
+					"some": "map",
+				},
+			},
+			src: &map[string]any{
+				"map": map[any]any{
+					"other": "map",
+				},
+			},
+			expected: &Example{
+				Map: map[any]any{
+					"other": "map",
+				},
+			},
+		},
+		{
+			name: "src(struct)",
+			dst:  &Example{},
+			src: &map[string]any{
+				"struct": map[string]any{
+					"int_field": 10,
+				},
+			},
+			expected: &Example{
+				Struct: Struct{
+					IntField: 10,
+				},
+			},
+		},
+		{
+			name: "dst(struct)",
+			dst: &Example{
+				Struct: Struct{
+					IntField: 10,
+				},
+			},
+			src: &map[string]any{},
+			expected: &Example{
+				Struct: Struct{
+					IntField: 10,
+				},
+			},
+		},
+		{
+			name: "overwrite(struct)",
+			dst: &Example{
+				Struct: Struct{
+					IntField: 10,
+				},
+			},
+			src: &map[string]any{
+				"struct": map[string]any{
+					"int_field": 20,
+				},
+			},
+			expected: &Example{
+				Struct: Struct{
+					IntField: 20,
+				},
+			},
+		},
+		{
+			name: "merge(struct)",
```

Comment:
Is this the behavior we want? We talked at some point about trying to behave like containerd where when you set a leaf node, it should overwrite the whole struct, not just the specific field. I'm not totally sold that's the right thing to do, but I'd like to understand what the tradeoffs are between the two and why we're going with this specific approach.